src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,372 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Base;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.BlockStore
+{
+    /// <summary>
+    /// Saves blocks to the database in batches, removes reorged blocks from the database.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// The batch is saved when total serialized size of all blocks in a batch reaches <see cref="BatchThresholdSizeBytes"/>,
+    /// or when more than <see cref="BatchMaxSaveIntervalSeconds"/> passed since last batch was saved, or when node is shutting down.
+    /// </para>
+    /// <para>
+    /// When we save new blocks to the database, in case <see cref="IBlockRepository"/> contains blocks that
+    /// are no longer a part of our best chain, they are removed from the database.
+    /// </para>
+    /// <para>
+    /// Block store is being initialized we delete blocks that are not on the best chain.
+    /// </para>
+    /// </remarks>
+    public class BlockStore : IDisposable
+    {
+        /// <summary>Maximum interval between saving batches.</summary>
+        private const int BatchMaxSaveIntervalSeconds = 37;
+
+        /// <summary>Maximum number of bytes the batch can hold until the downloaded blocks are stored to the disk.</summary>
+        internal const int BatchThresholdSizeBytes = 5 * 1000 * 1000;
```

Comment:
Do we want to express why theses numbers for the BatchThreshold and SaveInterval?

Reply:
Number for `BatchThresholdSizeBytes` came prom prev emplementation. Basically it's just a random reasonable number. 5mb sounds good for a batch size. 

37 is a prime number that wasn't used as an interval in any other component. Prime number is used here so we can avoid CPU consumption spikes (when node does nothing and at some point several events that consume CPU suddenly trigger).

Reply:
Ah that’s really interesting to use a prime number for that, I had never thought of that. So if you have many polling things using prime numbers as the interval, the clashes are less likely? If so, nice.

Reply:
I agree this knowledge could be added to the comment.

Reply:
this is a nice comment as it is good information.